content/learning-paths/smartphones-and-mobile/android_webgpu_dawn/5-render-a-simple-3D-object-part-1.md

```diff
@@ -0,0 +1,114 @@
+
+Here is a simplified diagram of WebGPU setup to draw triangles by using a vertex shader and a fragment shader:
+
+!["Triangle using WebGPU"](./images/webgpu-draw-high-level.svg "Figure 1. Triangle using WebGPU")
+
+The main things to notice in the above image are:
+
+* There is a **Pipeline**. It contains the vertex shader and fragment shader the GPU will run. You could also have a pipeline with a compute shader.
+* The shaders reference resources (buffers, textures, samplers) indirectly through **Bind Groups**.
+* The pipeline defines attributes that reference buffers indirectly through the internal state.
+* Attributes pull data out of buffers and feed the data into the vertex shader.
+* The vertex shader may feed data into the fragment shader.
+* The fragment shader writes to textures indirectly through the render pass description.
+
+To execute shaders on the GPU, you need to create all of these resources and set up this state. Creation of resources is relatively straightforward.
+
+{{% notice Note %}}
+Most WebGPU resources can not be changed after creation. You can change their contents but not their size, usage, format, etc… If you want to change any of that stuff you create a new resource and destroy the old one.
+{{% /notice %}}
+
+## Render Pipeline
+
+In order to achieve high performance real-time 3D rendering, the GPU processes shapes through a predefined pipeline. The pipeline itself is always the same but we can configure it in many ways. To do so, WebGPU provides a Render Pipeline object. The figure below illustrates the sequence of data processing stages executed by the render pipeline.
+
+!["Render Pipeline"](./images/render-pipeline.svg "Figure 2. Render Pipeline")
+
+Render Pipeline have to 2 main types of stages, **fixed-function** and **programmable**.
+
+### Fixed Functions stages
+
+Pipeline description consists in the following steps:
+
+* Describe vertex pipeline state
+* Describe vertex pipeline state
+* Describe primitive pipeline state
+* Describe fragment pipeline state
+* Describe stencil/depth pipeline state
+* Describe multi-sampling state
+* Describe pipeline layout
+
+The fixed function stages are well documented and you can refer to [code](https://github.com/varunchariArm/Android_DawnWebGPU/blob/main/app/src/main/cpp/webgpuRenderer.cpp#L256) and [further reading](https://eliemichel.github.io/LearnWebGPU/basic-3d-rendering/hello-triangle.html#lit-24) for configuring them. Configuring these stages is straight forward and is similar to other graphics APIs.
+
+### Programmable stage
+
+There are two programmable stages, namely give a vertex and a fragment shaders.
+
+### Shaders
+
+Both the vertex and fragment programmable stages use the same shader module, that we must first create. This module is kind of a dynamic library (like a .dll, .so or .dylib file), except that it talks the binary language of your GPU rather than your CPU’s.
```

Comment:
They *can* use the same shader module, but they don't have to.

Reply:
I see, updated